mesonbuild/backend/ninjabackend.py

```diff
@@ -1944,6 +1945,7 @@ def generate_link(self, target, outfile, outname, obj_list, linker, extra_args=[
         commands = []
         if not isinstance(target, build.StaticLibrary):
             commands += self.build.get_global_link_args(linker)
+            commands += self.build.get_project_link_args(linker, target.subproject)
```

Comment:
What we should be doing is gathering all the `-L` arguments from all sources into one `OrderedDict` and then generating the command from that. What you're pointing out is already a problem with our linker argument list since global link `-L` arguments override target-specific ones.